Add support for indented heredocs in perl

roflcopter4 opened the pull request:
Perl v5.26 added support for indented heredocs with the new operator '<<~', as in:
```
if ( something ) {
    my $foo = <<~ 'BAR';
    blah blah blah
    this $variable won't be syntax highlighted
    BAR
    if ( something further ) {
        my $baz = <<~ "QUX";
        but $this one will
        QUX
    }
}
```
At present vim syntax highlighting doesn't support this. In fact it gets rather confused. I tried to fix this in the simplest way possible, which was to allow the heredoc syntax highlighting to recognize either '<<' or '<<~' as valid operators, and to no longer require the end of document symbol to be at the start of the line. This seems to have worked. All the prove tests passed, and everything seems ok in a few canned scripts I wrote to test it out.
The only notable problem is that (Neo)Vim doesn't immediately switch the highlighting from 'string only' to an interpolated string (with variables highlighted) if the heredoc is suddenly changed from single to double quote, and vice versa. It does figure it out after a few moments though, and seems to have no problem normally - only when the type is changed after the fact.

roflcopter4 committed:
Add support for indented heredocs in perl

roflcopter4 committed:
Fix minor copy-paste error.

hoelzro commented:
Addresses #232

hoelzro commented:
Thanks @roflcopter4! I'll take a look at this

hoelzro commented:
Thanks again for the PR @roflcopter4  - unfortunately your code doesn't handle this highlighting properly:

```perl
do {
    print <<'EOF' if $true;
    Here document
    EOF
$true
EOF
}
```

...it improperly highlights the second occurence of `$true`, having assumed that the heredoc has ended.  I think what you submitted is a good start, though!

roflcopter4 committed:
Reimplement indented heredocs through duplication.

roflcopter4 committed:
Revert "Reimplement indented heredocs through duplication."

This reverts commit a5abedc1e1cf1767519948b44b8bf1763649f73e.

roflcopter4 committed:
Reimplement indented heredocs through duplication.

roflcopter4 commented:
I tried that piece of code in Neovim and it worked fine on my system. I was initially confused but I have (I think) figured out the problem. I didn't actually think this somewhat crude hack would be accepted, so I added a file to my after/syntax directory that contained a slightly modified form of this code. The version I uploaded definitely doesn't work - after removing that file it fails for me too.

The way I have it done in that file is about the crudest way possible. I copied the whole thing, added in the syntax for the indented heredoc, called it "region perlHereDoc2", and added the appropriate highlight link. Doing it this way does actually work. Defining the two forms as two totally different syntactic elements allows them both to be possible and unambiguous due to the different operator.

Now that I think about it, I'm not sure how you would get around that in just one go. I actually uploaded the code in this form, but duplicating it like that is pretty ugly.

hoelzro commented:
@roflcopter4 Unfortunately, I think that duplication is the only way to do it.  Thanks for doing the work to add this!
